Implement encoding interfaces for MAC address

onshorechet opened the pull request:
Implements the encoding TextMarshaler, BinaryMarshaler, TextAppender and BinaryAppender interfaces on the MAC type for efficient encoding. The String method is now written in terms of the new methods for consistency and improved performance.  This avoids unnecessary string allocations and conversions during the formatting process. This also allows MAC type to be used with other encoding packages.

onshorechet committed:
Add basic tests and benchmarks covering the mac.go file

onshorechet committed:
Implement encoding interfaces for MAC address

Implements the encoding TextMarshaler, BinaryMarshaler, TextAppender and BinaryAppender interfaces for MAC for efficient encoding of MAC addresses.
The String method is written in terms of the new methods for consistency and improved performance.  This avoids unnecessary string allocations and conversions during the formatting process.
This also allows MAC type to be used with other encoding packages like encoding/json, encoding/xml.

deadprogram commented:
Hello @onshorechet thanks for the PR, seems like a good idea.

Would you mind adding some test coverage for this? I realize we do not currently have a `mac_test.go` and seems like this PR would be very good place to add it. :smile_cat: 

Thank you!

onshorechet commented:
@deadprogram, I'm happy to!  Do you want me to close this PR and create a new one when the tests are ready?

deadprogram commented:
> Do you want me to close this PR and create a new one when the tests are ready?

You can just add another commit to this PR if you like, or rebase against this PR.

onshorechet commented:
After commit cce8bee the benchmark results are:
goos: linux
goarch: amd64
pkg: tinygo.org/x/bluetooth
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkParseMAC-12            47988309                24.78 ns/op
BenchmarkMacToString-12         80116750                13.75 ns/op
BenchmarkUUIDToString-12        11621157               113.2 ns/op
PASS
ok      tinygo.org/x/bluetooth  3.759s

onshorechet committed:
Refactored while writing tests.

onshorechet committed:
Add unsafe string cast that matches the spirit of what uuid is doing.

The uuid package is using the string builder package.  That package
works in bytes and self manages length.  Because mac address string
length is always know this commit uses the same method as string builder
when casting the bytes to string.

onshorechet commented:
Bench after 298fe32:
```
goos: linux
goarch: amd64
pkg: tinygo.org/x/bluetooth
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkParseMAC-12            44797298                25.16 ns/op
BenchmarkMacToString-12         83538506                13.22 ns/op
BenchmarkUUIDToString-12        13153808               109.5 ns/op
PASS
ok      tinygo.org/x/bluetooth  3.814s
```

onshorechet commented:
Final bench:
```
goos: linux
goarch: amd64
pkg: tinygo.org/x/bluetooth
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkParseMAC-12            45393645                24.35 ns/op
BenchmarkMacToString-12         100000000               10.16 ns/op
BenchmarkUUIDToString-12        13136982               102.2 ns/op
PASS
ok      tinygo.org/x/bluetooth  3.603s
```

onshorechet commented:
Sorry this got broken up into seperate commits but I waned to show the progression.  I wasn't sure where this project's priorities were where between speed, safety and readability.  In this usage of unsafe there doesn't seem to be a true safety problem it just becomes a bit harder to read.

I actually got this idea because I'm working on a similar changeset for the uuid file.  That file was already using strings.builder and the unsafe byte array to string cast is its secret sause.  In this case we know a mac address will always marshal to 17 bytes and the bytes are not kept or altered by anything else after the method returns.

onshorechet committed:
Add tests that cover MarshalText, AppendText, UnmarshalBinary and AppendBinary

deadprogram commented:
@onshorechet thanks for this work. So far in my testing everything is working and also no size changes.

@ysoldak or anyone else have any feedback before I squash/merge?

deadprogram commented:
OK now going to squash/merge. Thank you very much for working on this @onshorechet
